oap-server/server-storage-plugin/storage-jdbc-hikaricp-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/common/TableHelper.java

```diff
@@ -54,6 +58,16 @@ public class TableHelper {
     @Getter(lazy = true, value = AccessLevel.PRIVATE)
     private final ConfigService configService = moduleManager.find(CoreModule.NAME).provider().getService(ConfigService.class);
 
+    private final LoadingCache<String, Boolean> tableExistence =
+        CacheBuilder.newBuilder()
+                    .expireAfterWrite(10, TimeUnit.MINUTES)
+                    .build(new CacheLoader<>() {
+                        @Override
+                        public @NonNull Boolean load(@NonNull String tableName) throws Exception {
+                            return jdbcClient.tableExists(tableName);
+                        }
+                    });
```

Comment:
How about we have an `expireAfterRead` and do manually remove when `table not found` SQL exception? If possible, this could reduce more payload.

Reply:
> How about we have an `expireAfterRead` and do manually remove when `table not found` SQL exception? If possible, this could reduce more payload.

Hi, `expireAfterWrite` --> `expireAfterRead` sounds reasonable to me. However manually invalidating the cache doesn't make any sense to me, manually invalidating the cache only cause the `LoadingCache` to invoke the `load` method and tries to fill the cache again and thus increase the load, it **doesn't** sounds like "hey, if there is no entry in cache, don't do anything", this is how a loading cache works: load in the cache first, if absent, load from a remote source or load through a heavy computational operations, by manually invalidating the cache you are requesting the heavy computational operations over and over again because the cache entry is invalidated manually.

Reply:
Sounds like you want to capture the "table not found SQL exception", if that's the case, we don't even need a loading cache, we just use a plain cache (HashMap), and query the database for the first time, since then we only remove the cache when captured the exception, and the cache won't never expire unless exception is caught. Capturing such exception requires string manipulation though.

Reply:
> Sounds like you want to capture the "table not found SQL exception", if that's the case, we don't even need a loading cache, we just use a plain cache (HashMap), and query the database for the first time, since then we only remove the cache when captured the exception, and the cache won't never expire unless exception is caught. Capturing such exception requires string manipulation though.

Make sense. Map could be easier to archive this than cache.

Reply:
The cache structure is `<String table, Boolean exists>`, so if the `N - 7` tables are absent, the results will be cached for 10 minutes (`expireAfterWrite(10, TimeUnit.MINUTES)`), so within the 10 minutes, no query will be sent to database. It's nothing special for the `N - 7` tables actually.

Reply:
Let's just change to `expireAfterRead`. I think it is enough.

Reply:
Done